### PR TITLE
[SPARK-49050][SS] Integrate TransformWithState operator with Virtual Column Families

### PR DESCRIPTION
**Pull request by ericm-db**
### What changes were proposed in this pull request?

Persisting the mapping between columnFamilyName to columnFamilyId in RocksDBCheckpointMetadata and RocksDBSnapshot. On load RocksDB will propagate the mapping and the maximum column family ID so far to the provider, and the RocksDBStateStoreProvider will populate its mapping of columnFamilyName and columnFamilyId, and use this for any columnFamily operations. 
### Why are the changes needed?

To enable the use of virtual column families, and the performance benefits it comes along with, with the TransformWithState operator

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Amended unit tests

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala [hunk 1]
```diff
@@ -568,8 +577,8 @@ private[sql] class RocksDBStateStoreProvider
       if (!checkColFamilyExists(colFamilyName)) {
         val newColumnFamilyId = colFamilyId.incrementAndGet().toShort
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:578]: We can't generate the col family id on the executor. we need to do this on the driver

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SchemaHelper.scala [hunk 2]
```diff
@@ -36,7 +37,7 @@ case class StateSchemaFormatV3(
     stateStoreColFamilySchema: List[StateStoreColFamilySchema]
 )
 
-object SchemaHelper {
+object SchemaHelper extends Logging {
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SchemaHelper.scala:40]: do we need `Logging` here ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala [hunk 3]
```diff
@@ -379,6 +379,7 @@ trait StateStoreProvider {
       valueSchema: StructType,
       keyStateEncoderSpec: KeyStateEncoderSpec,
       useColumnFamilies: Boolean,
+      columnFamilyIds: Map[String, Short],
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala:382]: Could we just pass the array of the entire col family info here ?

**Comment by anishshri-db** [comment]
@ericm-db - test failures seem relevant ?

**Comment by ericm-db** [comment]
> @ericm-db - test failures seem relevant ?

It looks like the only failures are maintenance related?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala [hunk 4]
```diff
@@ -527,7 +546,7 @@ private[sql] class RocksDBStateStoreProvider
      * @param operationName - name of the store operation
      * @param colFamilyName - name of the column family
      */
-    private def verifyColFamilyCreationOrDeletion(
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:530]: intentional ?

**Comment by anishshri-db** [comment]
@ericm-db - could you update the PR description and describe all the changes this PR is doing ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala (hunk 4 above)
Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:530]: Yeah, so we could use it [here](https://github.com/apache/spark/pull/47524/files#diff-76a55d7a6f0b91a19366e01c919c615a4d13afc1c1c37b017afa9df1c1705218R72)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala [hunk 5]
```diff
@@ -166,21 +170,36 @@ class StateSchemaCompatibilityChecker(
   def validateAndMaybeEvolveStateSchema(
       newStateSchema: List[StateStoreColFamilySchema],
       ignoreValueSchema: Boolean,
-      stateSchemaVersion: Int): Boolean = {
+      stateSchemaVersion: Int): (Boolean, List[StateStoreColFamilySchema]) = {
     val existingStateSchemaList = getExistingKeyAndValueSchema().sortBy(_.colFamilyName)
-    val newStateSchemaList = newStateSchema.sortBy(_.colFamilyName)
+    val newStateSchemaList = newStateSchema.sortBy(_.colFamilyName).zipWithIndex.map {
```
Comment (ericm-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala:175]: Move inside the if condition

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala [hunk 6]
```diff
@@ -166,21 +170,36 @@ class StateSchemaCompatibilityChecker(
...
       // write the schema file if it doesn't exist
       createSchemaFile(newStateSchemaList, stateSchemaVersion)
-      true
+      (true, newStateSchemaList)
     } else {
       // validate if the new schema is compatible with the existing schema
-      existingStateSchemaList.lazyZip(newStateSchemaList).foreach {
-        case (existingStateSchema, newStateSchema) =>
-          check(existingStateSchema, newStateSchema, ignoreValueSchema)
+      val newList = existingStateSchemaList.map { existingSchema =>
```
Comment (ericm-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala:188]: newStateSchema.map

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala [hunk 7]
```diff
@@ -119,6 +119,8 @@ trait ReadStateStore {
  */
 trait StateStore extends ReadStateStore {
 
+  def columnFamilyIds: Map[String, Short] = Map.empty
```
Comment (ericm-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala:122]: getColumnFamilyIds

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala [hunk 8]
```diff
@@ -328,12 +335,19 @@ private[sql] class RocksDBStateStoreProvider
...
+    }.toMap
+  }
+
   override def init(
       stateStoreId: StateStoreId,
       keySchema: StructType,
       valueSchema: StructType,
       keyStateEncoderSpec: KeyStateEncoderSpec,
       useColumnFamilies: Boolean,
+      columnFamilySchemas: Map[String, StateStoreColFamilySchema],
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala:350]: Do we need this here ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImplBase.scala [hunk 9]
```diff
@@ -39,7 +39,7 @@ abstract class StatefulProcessorHandleImplBase(
         timeMode.toString)
     }
 
-    if (currState < INITIALIZED || currState >= TIMER_PROCESSED) {
+    if (currState < PRE_INIT || currState >= TIMER_PROCESSED) {
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImplBase.scala:42]: Is this change related ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 10]
```diff
@@ -703,6 +704,30 @@ object StreamExecution {
     "py4j.protocol.Py4JJavaError: An error occurred while calling" +
       s"((.|\\r\\n|\\r|\\n)*)(${IO_EXCEPTION_NAMES.mkString("|")})").r
 
+  private var columnFamilySchemas:
+    Map[UUID, Map[UUID, Map[String, StateStoreColFamilySchema]]] = Map.empty
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala:708]: I think we need to use a concurrent hashMap or use a mutex to make access to this map thread safe

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 11]
```diff
@@ -703,6 +704,30 @@ object StreamExecution {
...
+
+  def updateColumnFamilySchemas(
+      queryId: UUID,
+      runId: UUID,
+      newSchemas: Map[String, StateStoreColFamilySchema]): Unit = {
+    columnFamilySchemas = columnFamilySchemas.updated(
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala:714]: When/where are we removing entries from this map ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala [hunk 12]
```diff
@@ -577,9 +582,12 @@ case class TransformWithStateExec(
    */
   private def processData(store: StateStore, singleIterator: Iterator[InternalRow]):
     CompletionIterator[InternalRow, Iterator[InternalRow]] = {
+    val columnFamilyIds = getStateInfo.colFamilySchemas.map { case (name, schema) =>
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala:585]: Just move this to a common function ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala [hunk 13]
```diff
@@ -653,12 +665,32 @@ object TransformWithStateExec {
       initialStateDeserializer: Expression,
       initialState: SparkPlan): SparkPlan = {
     val shufflePartitions = child.session.sessionState.conf.numShufflePartitions
+
+    val driverProcessorHandle = new DriverStatefulProcessorHandleImpl(timeMode, keyEncoder)
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TransformWithStateExec.scala:669]: Lets add a comment explaining how this is for the batch world and why its safe

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala [hunk 14]
```diff
@@ -166,21 +170,37 @@ class StateSchemaCompatibilityChecker(
...
+      }
       createSchemaFile(newStateSchemaList, stateSchemaVersion)
-      true
+      (true, newStateSchemaList)
     } else {
       // validate if the new schema is compatible with the existing schema
-      existingStateSchemaList.lazyZip(newStateSchemaList).foreach {
-        case (existingStateSchema, newStateSchema) =>
-          check(existingStateSchema, newStateSchema, ignoreValueSchema)
+      val newList = newStateSchema.map { newSchema =>
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala:188]: Lets add a comment to explain what this is doing ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala [hunk 15]
```diff
@@ -166,21 +170,37 @@ class StateSchemaCompatibilityChecker(
...
+        existingStateSchemaList.find(_.colFamilyName == newSchema.colFamilyName) match {
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala:189]: Could we create a HashMap of the name->id mapping once outside and then do a O(1) lookup here for each col family ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala (hunk 10 above)
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala:708]: maybe we can just keep it simple and keep a map of `queryId -> ColFamilySchema` and update it on each new run and remove the entry on query stop ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala [hunk 16]
```diff
@@ -242,7 +242,26 @@ class IncrementalExecution(
...
+              schema.colFamilyName -> schema
+            }.toMap
+            StreamExecution.updateColumnFamilySchemas(
+              queryId, runId, columnFamilySchemas)
+            val stateInfo = tws.getStateInfo
+            tws.copy(stateInfo = Some(
+              stateInfo.copy(colFamilySchemas = columnFamilySchemas)))
+          case _ => statefulOp
+        }
+      case tws: TransformWithStateExec =>
```
[the remark this answers is not in the document]
Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala:259]: I assume we only want to populate for TWS, right? 
We need this block since everything above is only executed for the first batch of the run

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImplBase.scala (hunk 9 above)
Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImplBase.scala:42]: Yeah it is, since we call `registerTimer` in the PRE_INIT phase to add the columnFamilySchemas on the driver side.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala (hunk 16 above)
[the remark this answers is not in the document]
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala:259]: We already have a rule that populates the state info right ? could we just add to that ?

[the remark this answers is not in the document]
Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala:259]: Ah yeah we could

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImpl.scala [hunk 17]
```diff
@@ -121,12 +122,15 @@ class StatefulProcessorHandleImpl(
     metrics.get(metricName).foreach(_.add(1))
   }
 
+
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StatefulProcessorHandleImpl.scala:125]: nit: lets remove extra newline ?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateStoreColumnFamilySchemaUtils.scala [hunk 18]
```diff
@@ -16,49 +16,115 @@
...
 import org.apache.spark.sql.catalyst.encoders.ExpressionEncoder
+import org.apache.spark.sql.execution.streaming.StateTTLSchema.TTL_VALUE_ROW_SCHEMA
 import org.apache.spark.sql.execution.streaming.TransformWithStateKeyValueRowSchemaUtils._
-import org.apache.spark.sql.execution.streaming.state.{NoPrefixKeyStateEncoderSpec, PrefixKeyScanStateEncoderSpec, StateStoreColFamilySchema}
+import org.apache.spark.sql.execution.streaming.state.{NoPrefixKeyStateEncoderSpec, PrefixKeyScanStateEncoderSpec, RangeKeyScanStateEncoderSpec, StateStoreColFamilySchema}
+import org.apache.spark.sql.streaming.TimeMode
+import org.apache.spark.sql.types.{NullType, StructField, StructType}
 
-object StateStoreColumnFamilySchemaUtils {
+object StateStoreColumnFamilySchemaUtils extends Logging {
```
Comment (anishshri-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StateStoreColumnFamilySchemaUtils.scala:28]: nit: are we using `Logging` anywhere ?

**Comment by anishshri-db** [comment]
@ericm-db - are the docs failures related ?

**Review by anishshri-db — APPROVED** [review]
lgtm pending nits and green CI build

**Comment by ericm-db** [comment]
@HeartSaVioR PTAL when you get a chance

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StatefulProcessorHandleSuite.scala [hunk 19]
```diff
@@ -144,8 +147,16 @@ class StatefulProcessorHandleSuite extends StateVariableSuiteBase {
     test(s"registering timeouts with timeMode=$timeMode should succeed") {
       tryWithProviderResource(newStoreProviderWithStateVariable(true)) { provider =>
         val store = provider.getStore(0)
+        val timerCFName = if (timeMode == "ProcessingTime") {
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StatefulProcessorHandleSuite.scala:150]: Why not extracting this to the private method providing two CF names for given timeMode? Looks like we are introducing non-trivial number of duplicated code.

Comment (HeartSaVioR) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/StatefulProcessorHandleSuite.scala:150]: Maybe not a private method but utility object, as I'm seeing the same code in other suite as well.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ValueStateSuite.scala [hunk 20]
```diff
@@ -188,7 +192,8 @@ class ValueStateSuite extends StateVariableSuiteBase {
     val ex = intercept[StateStoreMultipleColumnFamiliesNotSupportedException] {
       provider.init(
         storeId, keySchema, valueSchema, NoPrefixKeyStateEncoderSpec(keySchema),
-        useColumnFamilies = true, storeConf, new Configuration)
+        useColumnFamilies = true,
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/ValueStateSuite.scala:195]: nit: unnecessary change?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala [hunk 21]
```diff
@@ -902,7 +902,8 @@ class TransformWithStateSuite extends StateStoreMetricsTest
           keySchema,
           new StructType().add("value", LongType, false),
           Some(NoPrefixKeyStateEncoderSpec(keySchema)),
-          None
+          None,
+          1.toShort
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateSuite.scala:906]: We predicted the future? :) The state names are alphabetically sorted.

**Review by HeartSaVioR — COMMENTED** [review]
First pass.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala [hunk 22]
```diff
@@ -166,21 +170,43 @@ class StateSchemaCompatibilityChecker(
...
-      existingStateSchemaList.lazyZip(newStateSchemaList).foreach {
-        case (existingStateSchema, newStateSchema) =>
-          check(existingStateSchema, newStateSchema, ignoreValueSchema)
+      // create a hashmap of name to schema from existing state schema list
+      val existingSchemas = existingStateSchemaList.map { schema =>
+        schema.colFamilyName -> schema
+      }.toMap
+      // For each column family we are creating, check if it existed in the
+      // previous run, and assign its ID to the old ID if the schema hasn't changed.
+      // If the schema has evolved, assign it a new ID.
```
[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala:193]: And I don't see we migrate the existing data, so this seems to be a hole of data loss.

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithValueStateTTLSuite.scala [hunk 23]
```diff
@@ -265,50 +265,129 @@ class TransformWithValueStateTTLSuite extends TransformWithStateTTLTest {
...
 
         val keySchema = new StructType().add("value", StringType)
+        val expirationSchema = new StructType().add("expirationMs", LongType)
+        val compositeKeySchema = new StructType()
+          .add("key", new StructType().add("value", StringType))
+          .add("userKey", new StructType()
+            .add("id", IntegerType, false)
+            .add("name", StringType))
+
+        // Define all 9 expected schemas
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithValueStateTTLSuite.scala:275]: Why not just define them as intuitive order (e.g. state schema and TTL schema closer) and assign CF ID later after sorting? It's probably easy to miss that the state names have to be sorted when we issue IDs for them.

**Comment by HeartSaVioR** [comment]
Also please update the PR description to contain the developer API change (StateStore interface) and persistent layer change (we store column family ID additionally). The code change seems to be a major one but PR description does not look like so.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala [hunk 24]
```diff
@@ -703,6 +706,24 @@ object StreamExecution {
     "py4j.protocol.Py4JJavaError: An error occurred while calling" +
       s"((.|\\r\\n|\\r|\\n)*)(${IO_EXCEPTION_NAMES.mkString("|")})").r
 
+  private val columnFamilySchemas:
```
[the remark this answers is not in the document]
Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala:709]: This isn't the StreamExecution class, this is the static object that is shared across multiple queries, right?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala [hunk 25]
```diff
@@ -166,21 +170,43 @@ class StateSchemaCompatibilityChecker(
   def validateAndMaybeEvolveStateSchema(
```
[the remark this answers is not in the document]
Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala:170]: Hm, we do check each state variable's schema, and evolve it individually in these [lines](https://github.com/apache/spark/pull/47524/files/808fa9c482789fd15447684928f5835bd3e06f5e#diff-978ab9a113cd63e6f5d8c1e5c98d0157c4127cfc4830a27cf5939260aaa783c7R194) right?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala [hunk 26]
```diff
@@ -129,6 +129,7 @@ trait StateStore extends ReadStateStore {
    */
   def createColFamilyIfAbsent(
       colFamilyName: String,
+      columnFamilyId: Short,
```
[the remark this answers is not in the document]
Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala:132]: We will deal with addition and removal in subsequent PRs.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala [hunk 27]
```diff
@@ -242,7 +243,20 @@ class IncrementalExecution(
             metadataWriter.write(metadata)
           case _ =>
         }
-        statefulOp
+        statefulOp match {
+          case tws: TransformWithStateExec =>
```
[the remark this answers is not in the document]
Comment (ericm-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala:247]: StatefulOperator doesn't have a copy method, which is why even the StateOpIdRule needs the specific class...

File: sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/MapStateSuite.scala [hunk 28]
```diff
@@ -234,7 +253,8 @@ class MapStateSuite extends StateVariableSuiteBase {
       val batchTimestampMs = 10
       val handle = new StatefulProcessorHandleImpl(store, UUID.randomUUID(),
         Encoders.STRING.asInstanceOf[ExpressionEncoder[Any]],
-        TimeMode.ProcessingTime(), batchTimestampMs = Some(batchTimestampMs))
+        TimeMode.ProcessingTime(), batchTimestampMs = Some(batchTimestampMs),
+        columnFamilyIds = Map("testState" -> 1.toShort))
```
[the remark this answers is not in the document]
Comment (ericm-db) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/streaming/state/MapStateSuite.scala:257]: Ah, I think because error is thrown before column family id verification

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala (hunk 24 above)
[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala:709]: I meant, we are creating each StreamExecution instance per query run, so if you scope this to the StreamExecution instance, you don't need a global cache at all. You don't have to refer the schema for query A from query B (cross reference), no?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala (hunk 27 above)
[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala:247]: Fair enough. Missed that.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala (hunk 25 above)
[the remark this answers is not in the document]
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateSchemaCompatibilityChecker.scala:173]: I meant, look at the code when we return true vs return false. We only return true when there wasn't schema information, so we don't really give the value whether it is really evolved or not.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala (hunk 24 above)
[the remark this answers is not in the document]
Comment (anishshri-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala:709]: Yea I think that should work - if there is an instance per query - that should also work I think
